Monarch Pro software integration tutorial (#33)
fictionlab/docs: docs/leo-rover/integrations/cameras/unispectral-monarch-pro.mdx

description: >-
  Learn how to connect a Monarch Pro multispectral camera to your Leo Rover for
  multispectral imaging.

image: /img/robots/leo/integrations/monarch-pro/monarch-pro.webp
---

This tutorial will guide you through the process of connecting a Monarch Pro
camera to your Leo Rover.

The [Monarch Pro](https://unispectral.com/monarch-pro/) is the ultimate camera
for advanced imaging, offering detailed material analysis through hyperspectral
imaging capabilities. It combines high-resolution spectral data capture with
real-time processing to provide comprehensive insights beyond the visible
spectrum. It can be used for material identification, crop health monitoring,
and industrial inspection. With its ability to capture and analyze a wide range
of wavelengths, the Monarch Pro enables precise analysis and decision-making
across multiple industries.

<ImageZoom
  src="/img/robots/leo/integrations/monarch-pro/monarch-pro.webp"
  width="1200"
  height="668"
  figStyle={{
    width: 600,
  }}
/>

## What to expect?

After completing this tutorial, you will have successfully connected a Monarch
Pro stereo camera to the Leo Rover in ROS, and you will possess basic knowledge

> about the use of multispectral camera.

Reviewer: ```suggestion
about the use of the multispectral camera.
```

Author: Done.

## Prerequisites

<LinkButton docId="leo-rover/guides/ssh" />
<LinkButton docId="leo-rover/guides/connect-to-network" />
<LinkButton docId="leo-rover/advanced-guides/install-ros-on-your-computer" />

## List of components

- Monarch Pro Multispectral Camera
- Any mounting solution
- USB-A to USB-C cable

## Software integration

We have prepared special ROS wrapper for the unispectral's Monarch Pro camera.

### Installing the packages

First you have to download it into `src` directory of your ROS workspace:

```bash
cd <path_to_your_ros_workspace>/src

> git clone https://github.com/fictionlab/ros_monarch_pro.git

Reviewer: What if the user does not have a directory for a local ROS workspace?

Author: Fair point - added a note about the need of having one and a link to our tutorial on how to create one.

```

> Having the source code, we need have installed all the required dependencies so
> the package will be installed successfully. To install the dependencies type:

Reviewer: ```suggestion
Having the source code, we need to install all of the required dependencies so
the package can be built successfully. To install the dependencies, type:
```

Author: Done.

```bash

> source devel/setup.bash
> rosdep install --from-paths src -iry

Reviewer: It's more failsafe to always update the package lists and rosdep database before running `rosdep install`
```suggestion
sudo apt update
rosdep update
rosdep install --from-paths src -iry
```

Reviewer: also, there is no `devel` directory if the workspace was never built previously

Author: Done - applied suggested changes and added info about the need of local ros workspace.

```

The node provides a ROS service, which saves a spectral cube into images to
specified directory. To trigger the image save you need to run in other terminal
(with your ROS workspace sourced):

```bash
rosservice call /monarch_pro_wrapper/save_images
```

If the save was successful you will receive the path to parent directory, where
the data is stored. Each call to the service results in creation of a separate
directory named like `cube_<date>` inside the parent directory, which contains
the saved files. By default the parent directory is set to your home directory.
To change it you can either change the argument inside the `wrapper.launch`
file, or you can set its value during launching the program - for example:

```bash
roslaunch monarch_ros wrapper.launch data_path:="/<your>/<desired>/<path>"
```

### Configuring the node

In the downloaded repository, there is a config file with camera parameters. The
parameters are loaded to the node automatically during the launch process. You
can change the provided camera parameters to your desired values (available
settings are `gain` and `exposure fps`). To load them into the program and your
camera, you need to restart the node.

> If you have ROS installed on your computer, you can also change the camera image
> with rqt plugin - dynamic reconfigure - which doesn't require node restart.

Reviewer: ```suggestion
If you have ROS installed on your computer, you can also change the camera wrapper parameters in runtime using the rqt dynamic reconfigure plugin.
```

Author: Done.

Before starting any of the programs, make sure you completed the **Connecting
another computer to ROS network** of **ROS development** tutorial:

<LinkButton docId="leo-rover/advanced-guides/ros-development" />

Now you can start `rqt`, and from `plugins` -> `configuration` choose
`dynamic reconfigure` and then choose the `monarch_pro_wrapper` node. On the
right side of the screen you will see the mentioned parameters - you can use the
sliders and boxes next to them to change the values of the parameters.

> 

Reviewer: Add a section about adding the node to autostart in `/etc/ros/robot.launch`

Author: I think it's redundant right now.
